[FEAT] Pre-Downloader window, Path Selection for install, New UX

DeerajTheepshi opened the pull request:
# What:
- A Pre Downloader window which lists the main packages that will be installed, allows the user to select the Path of installation.

# Why:
- Initially since we were installing a fixed set of packages, it din't matter to show the users what we were actually downloading, but now, since latest versions are being downloaded, it would be better to show it on a pre window.
- Also, as Crisitian suggested, it would be good to allow the users select their android sdk installation rather than using the sketchbookfolder. 

# How:
- There are a few methods added which facilitates the above.
- createInitLayout() : This method creates the initial pre window layout.
- addPackage(): This helps to add package to the window, reducing code redundancy.
- Start running the swing worker from createInitLayout(), when continue button is clicked. (Instead from the AndroidSDK Class)
- Query the repository , from the SDKDownloader constructor. **(NOTE: This is performed synchronously as of now, will later be converted to a swingWorker)**

# Todo:
- Convert synchronous task to swing worker
- Change dimensions as global variables as per java guidelines.

# Test:
- Similar to previous PR, the process can be initiated from the menu bar. 
- **Note: There will be a small delay due to the synchronous code**, and the window positioning is not solved yet. 
- 
![image](https://user-images.githubusercontent.com/5101112/58713495-52687080-83e0-11e9-98bf-749d1a62c369.png)

DeerajTheepshi committed:
[FIX] Download Links:Most Recent Version, Remove Redundant code

DeerajTheepshi committed:
[FEAT] Pre-Downloader window, Path Selection

kfrajer commented on mode/src/processing/mode/android/SDKDownloader.java:
```diff
@@ -192,22 +192,18 @@ protected Object doInBackground() throws Exception {
         Preferences.set("android.sdk.path", sdkFolder.getAbsolutePath());        
         sdk = AndroidSDK.load(false, null);        
-      } catch (ParserConfigurationException e) {
+      } catch (IOException e) {
```
I noticed you changed the multiple catches for a single catch. Can you comment in your decision? @DeerajTheepshi

DeerajTheepshi replied:
I haven't replaced them completely. All the removed exceptions, ie..ParserConfigurationException,SAXException, are thrown only by the XML parsing process that happens on getMainDownloadUrls() or getExtrasDownloadUrls(). Since they have been moved from the swing worker, these are not handled here. However you can find them at line 612 of the same class.

kfrajer commented on mode/src/processing/mode/android/SDKDownloader.java:
```diff
@@ -603,17 +599,35 @@ public static String humanReadableByteCount(long bytes, boolean si) {
+    downloadTask = new SDKDownloadTask(downloadUrls,sdkFolder);
     downloadTask.addPropertyChangeListener(this);
     downloadTask.execute();
-    setAlwaysOnTop(true);
```
I think these two lines could resolve the problem of modal window always on top: `setAlwaysOnTop(true);` and `setVisible(true);` Toggling these properties to false when window needs to be hidden. It is worth to test unless your research indicates there is a better way to do this. @DeerajTheepshi

DeerajTheepshi replied:
Yeah, this was my initial solution , but there is an issue with this. If you try implementing setAlwaysOnTop(false) to the dialog: I experienced this -> When switching to another window, say chrome, as expected the window is not visible. But now when I retry to open processing window via the task bar, processing PDE is not opened but only the dialog is opened. Only after closing the sdk download window, PDE becomes visible again . (Tested on : Win10 x86).  @kfrajer

kfrajer replied:
I see. Yes, this could be capture in the bug about modal window. This is important information and an interesting challenge to find the right solution, or the most convenient one. Feel free to mark as resolve. @DeerajTheepshi

kfrajer commented on mode/src/processing/mode/android/SDKDownloader.java:
```diff
@@ -623,6 +637,111 @@ public boolean cancelled() {
+    addPackage(gc,1,packagesPanel,"SDK PlatformTools: ",downloadUrls.platformToolsVersion);
+    addPackage(gc,2,packagesPanel,"Android Build Tools: ",downloadUrls.buildToolsVersion);
+    addPackage(gc,3,packagesPanel,"Android Tools: ",downloadUrls.toolsVersion);
+    addPackage(gc,4,packagesPanel,"Android Emulator: ",downloadUrls.emulatorVersion);
```
This indicates that during the installation process of the AndroidSDK, Processing will install the Android SDK as well as a standard emulator, correct? Potentially we could come back later to add extra details about the emulator. We can address this later after milestone 2 is done. Thoughts? @DeerajTheepshi

DeerajTheepshi replied:
I'm not sure about what you mean by adding extra details to emulator @kfrajer

kfrajer replied:
Sorry, I was not clear. I will send you a message in about 12hours about this. I would like to discuss about the user experience during the Android SDK installation as well as the emulator. Please mark this as resolved and i will start the discussion in the evol doc. @DeerajTheepshi

DeerajTheepshi committed:
[FIX] Current Directory in File chooser

DeerajTheepshi committed:
[FIX] Send querying to background

DeerajTheepshi commented:
@kfrajer , Added a few more commits. Do look into them

DeerajTheepshi committed:
[FIX] Update TARGET_SDK to Dowloaded platform

DeerajTheepshi committed:
[FIX] Remove Emulator Download

DeerajTheepshi commented:
@kfrajer , Added a few more commits, to hopefully conclude with the Automatic installer.

kfrajer commented on mode/src/processing/mode/android/SDKDownloader.java:
```diff
@@ -192,13 +192,14 @@ protected Object doInBackground() throws Exception {
          // Normalize built-tools and platform folders to android-<API LEVEL>
         String actualName = platformsFolder.listFiles()[0].getName();
-        renameFolder(platformsFolder, "android-" + TARGET_SDK, actualName);
+        renameFolder(platformsFolder, "android-" + downloadUrls.platformVersion, actualName);
```
Quick question. I notice you use downloadUrls to store information of downloaded components. I am considering to review this name and change it to a more descriptive name aligned to its content. What do you think @DeerajTheepshi?

DeerajTheepshi replied:
Yeah, A more specific name like packageHolder ? As its hold all information about packages? 
Also if we are changing this, downloadUrls belongs to class SDKUrlHolder, which can also be changed to packageManager or something similar?  @kfrajer

DeerajTheepshi committed:
[FIX]MacOS:FileDialog replace FileChooser

DeerajTheepshi committed:
[FIX]AlwaysOnTop Dialog,Add Back Option

DeerajTheepshi committed:
[Fix]Pre-Downloader UX changes

DeerajTheepshi commented:
@kfrajer , added a few more commits.

- Feature added: Back button for automatic installer.
- Feature added: Tooltip for path label.
- Feature added: JFileChooser doesn't work on Mac, so JDialog implemented
- Issues solved: #4 #3 
- Removed : Test code from editor.java file.

DeerajTheepshi committed:
[Fix]Error Handling,Remove test code

kfrajer commented:
@DeerajTheepshi, you can proceed with the PR. Note there are some conflicts that need to be resolved. Let me know if you want me to address them if you are not able to do it yourself.

DeerajTheepshi committed:
Merge branch 'android-0274-v4.1.0-beta3' into GSoC19-p1
